File: Python/hc.py
#Clusterin jerarquico

#importar librerias
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

#Cargar Dataser
dataset = pd.read_csv('Mall_Customers.csv')
X = dataset.iloc[:,[3,4]].values

#Dendrograma
import scipy.cluster.hierarchy as sch
dendrogram = sch.dendrogram(sch.linkage(X, method = "ward"))
plt.title("Dendrograma")
plt.xlabel("Clientes")
plt.ylabel("Distancia Euclidia")
plt.show()

#Ajustar Cluster jerarquico 
from sklearn.cluster import AgglomerativeClustering
hc = AgglomerativeClustering(n_clusters = 5 , affinity = "euclidean", linkage = "ward")
y_hc = hc.fit_predict(X)

# Visualizacion de Cluster
plt.scatter(X[y_hc == 0, 0], X[y_hc == 0, 1], s = 100, c = "red", label = "Cautos")
plt.scatter(X[y_hc == 1, 0], X[y_hc == 1, 1], s = 100, c = "blue", label = "Estandard")
plt.scatter(X[y_hc == 2, 0], X[y_hc == 2, 1], s = 100, c = "green", label = "Objetivo")
plt.scatter(X[y_hc == 3, 0], X[y_hc == 3, 1], s = 100, c = "cyan", label = "Descuidados")
plt.scatter(X[y_hc == 4, 0], X[y_hc == 4, 1], s = 100, c = "magenta", label = "Conservadores")
plt.title("Cluster de clientes")
plt.xlabel("Ingresos anuales (en miles de $)")
plt.ylabel("Puntuación de Gastos (1-100)")
plt.legend()
plt.show()

File: R/hc.R
#Clustering Jerarquico

#Cargar dataset
dataset = read.csv('Mall_Customers.csv')
X = dataset [,4:5]

#Dendograma
dendrogram = hclust(dist(X, method = "euclidean"),
                    method = "ward.D")
plot(dendrogram,
     main = "Dendograma",
     xlab = "Clientes del centro comercial",
     ylab = "Distancia Euclidea")

#Custering jerarquico
hc = hclust(dist(X, method = "euclidean"),
                    method = "ward.D")

y_hc = cutree(hc, k = 5)


#Visualizacion
library(cluster)
clusplot(X, 
         y_hc,
         lines = 0,
         shade = TRUE,
         color = TRUE,
         labels = 1,
         plotchar = FALSE,
         span = TRUE,
         main = "Clustering de clientes",
         xlab = "Ingresos anuales",
         ylab = "Puntuacion (1-100)"
)